### PR TITLE
Cronjob checker

### PR DESCRIPTION
**Pull request by jdowni000**
The cronjob-checker reaches out to a specified cronjob in a given namespace and lists events. It then ranges over the events for a specified reason. If there is an event with the specified reason in the given cronjob it will alert kuberhealthy. Issue #650

### CODE REVIEW

File: cmd/cronjob-checker/main.go [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+	}
+
+	//range over eventList for cronjob events that match provided reason
+	for _, e := range eventList.Items {
+		// log.Debugln(e.Reason)
+		if reason == e.Reason {
+			log.Infoln("There was an event with reason:" + e.Reason + " for cronjob" + cronJob)
+			reportErr := fmt.Errorf("cronjob: " + cronJob + " has an event with reason: " + reason)
+			ReportFailureAndExit(reportErr)
+		}
```
Comment (joshulyne) [inline cmd/cronjob-checker/main.go:78]: only question here, is how do we know the failure is coming from the cronjob? not sure if the cronjob name is being searched anywhere in the events, only the reason?

File: cmd/kuberhealthy/kuberhealthy.go [hunk 2]
```diff
@@ -1151,6 +1151,7 @@ func (k *Kuberhealthy) healthCheckHandler(w http.ResponseWriter, r *http.Request
 // this will return the state of ALL found checks.
 // Failures to fetch CRD state return an error.
 func (k *Kuberhealthy) getCurrentState(namespaces []string) health.State {
+	log.Debugln("getCurrentState: ", namespaces)
```
Comment (joshulyne) [inline cmd/kuberhealthy/kuberhealthy.go:1154]: is this intentional?

File: deploy/helm/kuberhealthy/templates/configmap.yaml [hunk 3]
```diff
@@ -28,6 +28,14 @@ data:
...
         annotations:
           description: Kuberhealthy shows that the cluster is not healthy
+      - alert: DeadMansSwitch
+        expr: vector(1)
+        labels:
+          severity: critical
+        annotations:
+          description: This is a DeadMansSwitch meant to ensure that the entire Alerting
+          pipeline is functional.
+        summary: Alerting DeadMansSwitch
```
Comment (joshulyne) [inline deploy/helm/kuberhealthy/templates/configmap.yaml:38]: curious what this is for?

**Comment by joshulyne** [comment]
just a few comments! wondering if we can expand this to check on all cronjobs in a given namespace maybe? instead of just checking one cronjob at a time?

File: cmd/kuberhealthy/kuberhealthy.go (hunk 2 above)
Comment (jdowni000) [reply cmd/kuberhealthy/kuberhealthy.go:1154]: this was not...perhaps Eric and I wrote this when we were debugging something earlier...have no idea how that got in there

File: deploy/helm/kuberhealthy/templates/configmap.yaml (hunk 3 above)
Comment (jdowni000) [reply deploy/helm/kuberhealthy/templates/configmap.yaml:38]: these must be coming from merging...although I thought this was only master that I had been doing it too....I am not sure but will get rid of it

**Comment by integrii** [comment]
- Include the cronjob name that is failing in the kuberhealthy error report
- remove the reason env var and hard-code it into place (for now)
- consider removing namespace env var reference in the readme, because it is automatically set to the namespace the check gets deployed into

File: cmd/cronjob-checker/README.md [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+  runInterval: 1h
+  timeout: 10m
+  podSpec:
+    serviceAccountName: cronjob-checker
+    containers:
+      - name: cronjob-checker
+        image: kuberhealthy/cronjob-checker:v1.1.0
+        imagePullPolicy: IfNotPresent
+        env:
+          - name: NAMESPACE
```
Comment (integrii) [inline cmd/cronjob-checker/README.md:36]: True this up with the `fieldRef` in your example yaml.

**Review by integrii — CHANGES_REQUESTED** [review]

File: cmd/cronjob-checker/README.md (hunk 4 above)
Comment (jdowni000) [reply cmd/cronjob-checker/README.md:32]: done

**Comment by integrii** [comment]
- remove all namespaces from flag spec files in readme
- add specs to helm chart so that this deploys into the helm-specified namespace
- change clusterrolebindings to rolebindings

File: cmd/cronjob-checker/main.go (hunk 1 above)
Comment (joshulyne) [reply cmd/cronjob-checker/main.go:69]: if theres one event failure, it reports failure and the program exits... what if there are multiple cronjob failures? i think we should be able to catch them all 😏

File: cmd/cronjob-checker/README.md [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+      - "events.k8s.io"
+    resources:
+      - events
+    verbs:
+      - get
+      - list
+      - watch
+---
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRoleBinding
```
Comment (joshulyne) [inline cmd/cronjob-checker/README.md:55]: the example uses `ClusterRoleBinding` while the `cmd/cronjob-checker.yaml` uses `RoleBinding`

File: cmd/cronjob-checker/cronjob-checker.yaml [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+      - "events.k8s.io"
+    resources:
+      - events
+    verbs:
+      - get
+      - list
+      - watch
+---
+apiVersion: rbac.authorization.k8s.io/v1
+kind: RoleBinding
```
Comment (joshulyne) [inline cmd/cronjob-checker/cronjob-checker.yaml:44]: should this be `ClusterRoleBinding`?

File: cmd/cronjob-checker/README.md (hunk 5 above)
Comment (jdowni000) [reply cmd/cronjob-checker/README.md:55]: I have moved this to a Role and Role Binding

File: cmd/cronjob-checker/cronjob-checker.yaml (hunk 6 above)
Comment (jdowni000) [reply cmd/cronjob-checker/cronjob-checker.yaml:44]: I have moved this to a Role and Role Binding

File: go.sum [hunk 7]
```diff
@@ -392,15 +392,23 @@ honnef.co/go/tools v0.0.0-20190523083050-ea95bdfd59fc/go.mod h1:rf3lG4BRIbNafJWh
 honnef.co/go/tools v0.0.1-2019.2.3/go.mod h1:a3bituU0lyd329TUQxRnasdCoJDkEUEAqEt0JzvZhAg=
 k8s.io/api v0.19.0 h1:XyrFIJqTYZJ2DU7FBE/bSPz7b1HvbVBuBf07oeo6eTc=
 k8s.io/api v0.19.0/go.mod h1:I1K45XlvTrDjmj5LoM5LuP/KYrhWbjUKT/SoPG0qTjw=
+<<<<<<< HEAD
```
Comment (integrii) [inline go.sum:395]: remove these lines to resolve the conflict markers pls

**Review by integrii — CHANGES_REQUESTED** [review]

File: go.sum (hunk 7 above)
Comment (jdowni000) [reply go.sum]: I have removed 392-394

**Review by integrii — CHANGES_REQUESTED** [review]
just a few small changes - otherwise this LGTM.  We can circle around and make `findLastCronRunTime` more reasonable later.

File: cmd/cronjob-checker/README.md [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+  name: cronjob-checker
+spec:
+  runInterval: 5m
+  timeout: 10m
+  podSpec:
+    serviceAccountName: cronjob-checker
+    containers:
+      - name: cronjob-checker
+        image: kuberhealthy/cronjob-checker:v2.0.0
+        imagePullPolicy: Always
```
[the remark this answers is not in the document]
Comment (jdowni000) [reply cmd/cronjob-checker/README.md:21]: done

File: cmd/cronjob-checker/cronjob-checker.yaml [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+  name: cronjob-checker
+spec:
+  runInterval: 5m
+  timeout: 10m
+  podSpec:
+    serviceAccountName: cronjob-checker
+    containers:
+      - name: cronjob-checker
+        image: kuberhealthy/cronjob-checker:v2.0.0
+        imagePullPolicy: Always
```
[the remark this answers is not in the document]
Comment (jdowni000) [reply cmd/cronjob-checker/cronjob-checker.yaml:13]: done

File: cmd/cronjob-checker/main.go [hunk 10]
```diff
@@ -22,16 +22,25 @@ var kubeConfigFile = filepath.Join(os.Getenv("HOME"), ".kube", "config")
...
+		log.Fatalln("Error parsing KH_CHECK_RUN_DEADLINE:", err)
+	}
+	khCheckRunDeadline := time.Unix(intkhCheckRundDeadline, 0)
+
+	khCheckRunTime := khCheckRunDeadline.Sub(time.Now()) / 2
 
-	// create context
-	checkTimeLimit := time.Minute * 1
-	ctx, _ := context.WithTimeout(context.Background(), checkTimeLimit)
+	log.Infoln(khCheckRunDeadline)
```
Comment (integrii) [inline cmd/cronjob-checker/main.go:42]: For production code, let's either drop this log line.  Also, leaving a comment on the check run deadline code above would be nice.

File: cmd/cronjob-checker/main.go [hunk 11]
```diff
@@ -81,12 +90,11 @@ func main() {
 		lastRunTimeV1 := cronGet.Status.LastScheduleTime
 		lastRunTime := lastRunTimeV1.Time
 		shouldOfRun := findLastCronRunTime(schedule)
-		timeMinus := shouldOfRun.Add(time.Duration(-30) * time.Second)
-		timePlus := shouldOfRun.Add(time.Duration(30) * time.Second)
+		tM, tP := scheduleWindow(shouldOfRun, time.Minute*5)
```
Comment (integrii) [inline cmd/cronjob-checker/main.go:93]: `tM` and `tP` are a little ambiguous here.  Could we rename this to something like `earliestRunTime` and `latestRunTime`

File: cmd/cronjob-checker/main.go [hunk 12]
```diff
@@ -81,12 +90,11 @@ func main() {
...
 
 		log.Infoln("last run time for cronjob", c.Name, "was", lastRunTime)
 
-		if lastRunTime.Before(timeMinus) && lastRunTime.Before(timePlus) {
+		if lastRunTime.Before(tM) && lastRunTime.Before(tP) {
```
Comment (integrii) [inline cmd/cronjob-checker/main.go:97]:
```suggestion
		if lastRunTime.After(tM) && lastRunTime.Before(tP) {
```

What we want to say here is: "if the last run time was after the `earliestAllowedTime` and the last run time was before the `latestAllowedTime`.

**Review by integrii — CHANGES_REQUESTED** [review]

File: cmd/cronjob-checker/main.go (hunk 10 above)
Comment (jdowni000) [reply cmd/cronjob-checker/main.go:42]: removed and comment added

File: cmd/cronjob-checker/main.go (hunk 11 above)
Comment (jdowni000) [reply cmd/cronjob-checker/main.go:93]: done

File: cmd/cronjob-checker/main.go (hunk 12 above)
Comment (jdowni000) [reply cmd/cronjob-checker/main.go:97]: done. changed logic to make this the positive

**Review by integrii — APPROVED** [review]
Looks great!  Thanks!

**Pull request merged**